src/Init/Data/AC.lean

```diff
@@ -0,0 +1,328 @@
+    simp [evalList_insert ctx h]
+    cases h₂ : insert y xs
+    . apply absurd h₂; simp [insert_nonEmpty]
+    . simp [evalList, ←h₂, evalList_insert ctx h]
+    all_goals simp [insert_nonEmpty]
+
+theorem Context.evalList_sort_cons
+  (ctx : Context α)
+  (h : IsCommutative ctx.op)
+  (x : Nat)
+  (xs : List Nat)
+  : evalList α ctx (sort (x :: xs)) = evalList α ctx (x :: sort xs) := by
+  simp [sort, sort.loop]
+  generalize [] = ys
+  induction xs generalizing x ys with
+  | nil => simp [sort.loop, evalList_insert ctx h]
+  | cons z zs ih =>
+    rw [evalList_sort_loop_swap ctx h]; simp [sort.loop, ←ih]; apply evalList_sort_congr ctx h; rw [evalList_insert ctx h]
+    cases h₂ : insert x ys with
+    | nil => apply absurd h₂; simp [insert_nonEmpty]
+    | cons u us =>
+      cases h₃ : insert z ys with
+      | nil => apply absurd h₃; simp [insert_nonEmpty]
+      | cons v vs =>
+        simp [evalList, ←h₂, ←h₃, evalList_insert ctx h]
+        cases ys
+        . simp [evalList, h.1, EvalInformation.evalOp]
+        . simp [evalList, EvalInformation.evalOp]; rw [h.1, ctx.assoc.1, h.1 (evalList _ _ _)]
+    all_goals simp [insert_nonEmpty]
+
+theorem Context.evalList_sort (ctx : Context α) (h : ContextInformation.isComm ctx) (e : List Nat) : evalList α ctx (sort e) = evalList α ctx e := by
+  have h : IsCommutative ctx.op := by simp [ContextInformation.isComm, Option.isSome] at h; cases h₂ : ctx.comm <;> simp [h₂] at h; assumption
+  induction e using List.two_step_induction with
+  | empty => rfl
+  | single => rfl
+  | step x y ys ih =>
+    simp [evalList_sort_cons ctx h]
+    cases h₂ : sort (y :: ys) with
+    | nil => simp [sort, sort.loop] at *; apply absurd h₂; apply sort_loop_nonEmpty; apply insert_nonEmpty
+    | cons z zs => simp [evalList, ←h₂, ih]
+
+theorem Context.toList_nonEmpty (e : Expr) : e.toList ≠ [] := by
+  induction e with
+  | var => simp [Expr.toList]
+  | op l r ih₁ ih₂ =>
+    simp [Expr.toList]
+    cases h : l.toList with
+    | nil => contradiction
+    | cons => simp [List.append]
+
+theorem Context.unwrap_isNeutral
+  {ctx : Context α}
+  {x : Nat}
+  : ContextInformation.isNeutral ctx x = true → IsNeutral (EvalInformation.evalOp ctx) (EvalInformation.evalVar (β := α) ctx x) := by
+  simp [ContextInformation.isNeutral, Option.isSome, EvalInformation.evalOp, EvalInformation.evalVar]
+  match (var ctx x).neutral with
+  | some hn => intro; assumption
+  | none => intro; contradiction
+
+theorem Context.evalList_removeNeutrals (ctx : Context α) (e : List Nat) : evalList α ctx (removeNeutrals ctx e) = evalList α ctx e := by
+  induction e using List.two_step_induction with
+  | empty => rfl
+  | single =>
+    simp [removeNeutrals, removeNeutrals.loop]; split
+    case h_1 => rfl
+    case h_2 h _ _ => split at h <;> simp_all
+  | step x y ys ih =>
+    match h₁ : ContextInformation.isNeutral ctx x, h₂ : ContextInformation.isNeutral ctx y, h₃ : removeNeutrals.loop ctx ys with
```

Comment:
We can use `<;>` to avoid this kind of big `match`. 
```lean
    cases h₁ : ContextInformation.isNeutral ctx x <;> cases h₂ : ContextInformation.isNeutral ctx y <;> cases h₃ : removeNeutrals.loop ctx ys 
      <;> simp [removeNeutrals, removeNeutrals.loop, h₁, h₂, h₃, evalList, ←ih] <;> (try simp [unwrap_isNeutral h₂ |>.2])
      <;> (try simp [unwrap_isNeutral h₁ |>.1])
```

Reply:
done